File: mobile/lib/services/itinerary_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../core/constants/api_constants.dart';
import '../models/itinerary_generated_model.dart';
import '../models/itinerary_detail_model.dart';
import '../models/riwayat_model.dart';
import 'api_service.dart';

class ItineraryService {
  final Dio _dio;

  ItineraryService(this._dio);

  Future<ItineraryGeneratedModel> generateItinerary({
    required int totalBudget,
    required int durationDays,
    required double startLatitude,
    required double startLongitude,
    String? cityPreference,
    String? categoryPreference,
  }) async {
    try {
      final payload = {
        'total_budget': totalBudget,
        'duration_days': durationDays,
        'start_latitude': startLatitude,
        'start_longitude': startLongitude,
        if (cityPreference != null && cityPreference.isNotEmpty)
          'city_preference': cityPreference,
        if (categoryPreference != null && categoryPreference.isNotEmpty)
          'category_preference': categoryPreference,
      };

      final response = await _dio.post(ApiConstants.generateItinerary, data: payload);
      return ItineraryGeneratedModel.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      throw _pesanError(e, 'Gagal membuat itinerary. Coba lagi nanti.');
    }
  }

  Future<ItineraryDetailModel> getItineraryById(int id) async {
    try {
      final response = await _dio.get('${ApiConstants.itinerary}/$id');
      return ItineraryDetailModel.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      throw _pesanError(e, 'Gagal memuat detail itinerary.');
    }
  }

  Future<List<RiwayatItemModel>> getRiwayat() async {
    try {
      final response = await _dio.get(ApiConstants.userItineraries);
      final data = (response.data as Map<String, dynamic>)['data'] as List<dynamic>;
      return data
          .map((item) => RiwayatItemModel.fromJson(item as Map<String, dynamic>))
          .toList();
    } on DioException catch (e) {
      throw _pesanError(e, 'Gagal memuat riwayat itinerary.');
    }
  }

  Future<void> deleteItinerary(int id) async {
    try {
      await _dio.delete('${ApiConstants.itinerary}/$id');
    } on DioException catch (e) {
      throw _pesanError(e, 'Gagal menghapus itinerary.');
    }
  }

  Future<List<String>> getDaftarKota() async {
    try {
      final response = await _dio.get(ApiConstants.destinationCities);
      final data = (response.data as Map<String, dynamic>)['data'] as List<dynamic>;
      return data.map((k) => k as String).toList();
    } on DioException catch (e) {
      throw _pesanError(e, 'Gagal memuat daftar kota.');
    }
  }

  String _pesanError(DioException e, String fallback) {
    final data = e.response?.data;
    if (data is Map<String, dynamic>) {
      return data['pesan'] as String? ?? fallback;
    }
    if (e.type == DioExceptionType.connectionTimeout ||
        e.type == DioExceptionType.receiveTimeout) {
      return 'Koneksi ke server timeout. Pastikan server berjalan.';
    }
    if (e.type == DioExceptionType.connectionError) {
      return 'Tidak dapat terhubung ke server. Periksa koneksi jaringan.';
    }
    return fallback;
  }
}

final itineraryServiceProvider = Provider<ItineraryService>((ref) {
  return ItineraryService(ref.read(dioProvider));
});

File: mobile-rn/src/services/itineraryService.js
import api from './api';

function pesanError(error, fallback) {
  const data = error.response?.data;
  if (data?.pesan) return data.pesan;
  if (error.code === 'ECONNABORTED') return 'Koneksi timeout. Pastikan server berjalan.';
  if (error.message === 'Network Error') return 'Tidak dapat terhubung ke server.';
  return fallback;
}

export async function generateItinerary({ duration, budget, startLat, startLng, city, category }) {
  try {
    const payload = {
      duration_days: duration,
      total_budget: budget,
      start_latitude: startLat,
      start_longitude: startLng,
      ...(city && { city_preference: city }),
      ...(category && { category_preference: category }),
    };
    const res = await api.post('/api/itinerary/generate', payload);
    return res.data;
  } catch (error) {
    throw pesanError(error, 'Gagal membuat itinerary. Coba lagi nanti.');
  }
}

export async function getItinerary(id) {
  try {
    const res = await api.get(`/api/itinerary/${id}`);
    return res.data;
  } catch (error) {
    throw pesanError(error, 'Gagal memuat detail itinerary.');
  }
}

export async function getRiwayat() {
  try {
    const res = await api.get('/api/itinerary/riwayat');
    return res.data.data;
  } catch (error) {
    throw pesanError(error, 'Gagal memuat riwayat perjalanan.');
  }
}

export async function deleteItinerary(id) {
  try {
    await api.delete(`/api/itinerary/${id}`);
  } catch (error) {
    throw pesanError(error, 'Gagal menghapus itinerary.');
  }
}

export async function getDaftarKota() {
  try {
    const res = await api.get('/api/destinasi/kota');
    return res.data.data;
  } catch (error) {
    return [];
  }
}
